File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        if (!root) {
            return nullptr;
        }
        queue<TreeNode*> q;
        q.push(root);
        vector<int> levelSums;
        while (!q.empty()) {
            int size = q.size();
            int levelSum = 0;
            for (int i = 0; i < size; ++i) {
                TreeNode* node = q.front();
                q.pop();
                levelSum += node->val;
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            levelSums.push_back(levelSum);
        }
        root->val = 0;
        q.push(root);
        int level = 1;
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                TreeNode* node = q.front();
                q.pop();
                int sumChildren = 0;
                if (node->left) {
                    sumChildren += node->left->val;
                }
                if (node->right) {
                    sumChildren += node->right->val;
                }
                if (node->left) {
                    node->left->val = levelSums[level] - sumChildren;
                    q.push(node->left);
                }
                if (node->right) {
                    node->right->val = levelSums[level] - sumChildren;
                    q.push(node->right);
                }
            }
            level++;
        }
        return root;
    }
};

File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        q=[]
        q.append(root)
        ls=[]
        while(len(q)!=0):
            s=len(q)
            sum=0
            for i in range(s):
                n=q.pop(0)
                sum+=n.val
                if n.left:
                    q.append(n.left)
                if n.right:
                    q.append(n.right)
            ls.append(sum)
        root.val=0
        q.append(root)
        level=1
        while(len(q)!=0):
            s=len(q)
            for i in range(s):
                n=q.pop(0)
                sum=0
                if n.left:
                    sum+=n.left.val
                if n.right:
                    sum+=n.right.val
                if n.left:
                    n.left.val=ls[level]-sum
                    q.append(n.left)
                if n.right:
                    n.right.val=ls[level]-sum
                    q.append(n.right)
            level+=1
        return root
        

        